src/proposed_new_lib_api.rs

```diff
@@ -0,0 +1,190 @@
+    /// For example, perhaps the package does not yet exist at the specified revision.
+    if_failed_to_find_package: OnUnexpectedOutcome,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ExactRegistryVersion {
+    /// The crate version to check.
+    version: semver::Version,
+
+    /// If we aren't able to use the registry, for example due to a connection error.
+    if_failed_to_use_registry: OnUnexpectedOutcome,
+
+    /// If we connected to the registry but could not find the specified package.
+    if_failed_to_find_package: OnUnexpectedOutcome,
+
+    /// If the registry knows about the package, but does not have the specified version.
+    if_failed_to_find_package_version: OnUnexpectedOutcome,
+
+    /// If the registry has the specified version but it is not usable for semver-checking:
+    /// for example, if it has been yanked (we won't be able to generate rustdoc for it).
+    if_version_is_unusable: OnUnexpectedOutcome,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PredicateRegistryVersion {
+    /// The way to select the version.
+    predicate: VersionPredicate,
+
+    /// Whether to consider pre-release versions or not.
+    ///
+    /// Note that breaking changes are always allowed between pre-releases,
+    /// and between a pre-release and a corresponding normal release.
+    allow_prereleases: bool,
+
+    /// Whether we should automatically ignore yanked releases.
+    ignore_yanked: bool,
+
+    /// If we aren't able to use the registry, for example due to a connection error.
+    if_failed_to_use_registry: OnUnexpectedOutcome,
+
+    /// If we connected to the registry but could not find the specified package.
+    if_failed_to_find_package: OnUnexpectedOutcome,
+
+    /// If the registry knows about the package, but none of its versions match our requirements.
+    if_no_matching_version: OnUnexpectedOutcome,
+
+    /// If a version matched our requirements, but it is not usable for semver-checking:
+    /// for example, if it has been yanked (we won't be able to generate rustdoc for it).
+    if_version_is_unusable: OnUnexpectedOutcome,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VersionPredicate {
+    /// Strictly less than the specified version.
+    LessThan(semver::Version),
+
+    /// Less than or equal than the specified version.
+    LessThanOrEqual(semver::Version),
+
+    /// The largest version number available.
+    HighestAvailable,
+
+    /// The last release ordered by release date.
+    MostRecentlyPublished.
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum OnUnexpectedOutcome {
+    /// Stop running checks and return an error.
+    FailStop,
+
+    /// Log an error but continue with other checks.
+    LogAndContinue,
+
+    /// Continue with other checks as if nothing happened.
+    /// Do not print anything to stderr.
+    ContinueSilently,
+}
```

Comment:
An alternative to this design would be to model the internals of `cargo semver-checks` as a state machine that a binary or other consumer needs to manually advance. That is a bit more work to integrate but gives you greater flexibility. In other words, the library part would only offer building blocks that you can compose together with your own policy.

For example, if said state machine yields an event like:

```rust
struct WorkspaceCrateDiscovered {
	pub publish: bool,
	pub name: String
	// ...
}
```

We would also need APIs like:

```rust
pub fn load_crate_from_registry(name: &str) -> Result<...> { }

pub fn check_semver(base: &CrateData, new: &CrateData) -> Result<impl Iterator<Item = SemverViolation>, ...> {}
```

Upon encountering a `WorkspaceCrateDiscovered` event a user could then e.g. decide whether or not to continue checking the crate for semver-violations based on the `publish` field.